### ISSUE TITLE
difference between fasttext measure and other measures

### ISSUE THREAD

**Issue by koenvanderveen** [facebookresearch/fastText#151]
When i played around with fasttext for a bit, i noticed that the scores returned by model.test (in python) are different from the scores i computed myself. I compared the p@1 and r@1 from model.test with the accuracy i computed myself based on the predictions of model.predict. My own computations give me values around  0.80 for ag_news while i get 0.92 using model.test. I think accuracy and P@1 or R@1 must be the same since i use k=1. Just to make sure i tried the sklearn.metrics accuracy_score aswell, which also gave me 0.80.

Am I missing something obvious here?

**Comment by denimboy** [comment]
Hi,

I am also having some trouble understanding these measures.  I think it's precision @ K and recall @ K?  If P@K is the "number of times the correct label was seen in the top K results."  Not sure what recall @ K means but perhaps its "number of times correct label was returned as a potential answer even after K"?

I see the relevant lines in fasttext.cc:

`std::cout << "P@" << k << ": " << precision / (k * nexamples) << std::endl;`
` std::cout << "R@" << k << ": " << precision / nlabels << std::endl;`

**precision** gets incremented by 1.0 when the correct label is seen in the list of K labels returned by predict().  **nexamples** is incremented by one for each test case.  **nlabels** gets incremented by the number of labels in the result set which is usually just K.

In my case I had text with 180 potential labels.  Each example gets one and only one label.  I was trying to classify and get scores for K in 1, 5, and 10.  To my surprise p@K went down as K went up.  r@K went up as K increased.

My understanding now is that r@k is the percentage of examples the correct label was seen in the top K.  p@k really is the precision at K and thus is the percentage of correct labels seen in the top K.  This kind of infers that fasttext can learn multiple labels but I have not tried this.

A little clarification on what to make of these metrics from the facebook team would be helpful.

Thanks,

**Comment by bkj** [comment]
P@k is percentage of labels in top k that are correct. R@k is percentage of
correct labels that are returned in the top k.

Fasttext can indeed learn multiple labels -- at each pass it randomly
samples one of the labels, so after multiple epochs a given X is "pushed"
towards each label an equal number of times.

**Comment by koenvanderveen** [comment]
I think i understand p@k and r@k correctly. I think p@1 and r@1 should be equivalent to accuracy for classfication tasks (such as agnews) where there is only 1 correct label for every datapoint. However, when I compare these measures with the accuracy computed from the predictions there is a difference

**Comment by CARTON-BOX** [comment]
ha ha

**Comment by pythonner** [comment]
We had the same problem when testing on a validation set with some unknown labels, i.e., labels that were not seen during training. Looks like Fasttext isn't penalizing for not retrieving these unknown labels and, therefore, R@k computed by Fasttext are higher than what it should.

Similarly, if a test line only contains unknown labels, it won't be taken into account by the test routine. The 'Number of examples' [counted by the test](https://github.com/facebookresearch/fastText/blob/master/src/fasttext.cc#L335) won't include the said line. Again, this artificially increases final R@k value.

**Comment by matanster** [comment]
I think you should anyway use an external tool (or roll your own) for obtaining your ultimate measure. There's just too many measures to choose from, and you want to make sure your measure is tuned to the measure just-right for your use case.
